gto-support-picasso/src/main/java/org/ccci/gto/android/common/picasso/RequestCreatorCoroutines.kt

```diff
@@ -5,17 +5,23 @@ import android.graphics.drawable.Drawable
 import com.squareup.picasso.Picasso
 import com.squareup.picasso.RequestCreator
 import com.squareup.picasso.Target
-import kotlin.coroutines.Continuation
+import com.squareup.picasso.picasso
 import kotlin.coroutines.resume
 import kotlin.coroutines.resumeWithException
-import kotlin.coroutines.suspendCoroutine
+import kotlinx.coroutines.CancellableContinuation
 import kotlinx.coroutines.Dispatchers
+import kotlinx.coroutines.suspendCancellableCoroutine
 import kotlinx.coroutines.withContext
 
-suspend fun RequestCreator.getBitmap(): Bitmap =
-    withContext(Dispatchers.Main) { suspendCoroutine { into(BitmapContinuationTarget(it)) } }
+suspend fun RequestCreator.getBitmap(): Bitmap = withContext(Dispatchers.Main) {
+    suspendCancellableCoroutine { cont ->
+        val target = BitmapContinuationTarget(cont)
+        into(target)
+        cont.invokeOnCancellation { picasso?.cancelRequest(target) }
+    }
+}
```

Comment:
@frett  I think I understand what's going on here, but could you explain this just to make sure I have it right? Is this ```target``` the strong reference you created and what does the ```into(target)``` do with that reference?

Reply:
now that we've had the memory conversation, before we were creating an object (`BitmapContinuationTarget(cont)`) which we sent into the `into()` method, `into()` queued up a background process to execute and immediately returned. This caused the local reference to the target to go out of scope, and because `into` only held a weak reference the target was available for garbage collection.

the current code maintains a strong reference to the target via the `invokeOnCancellation` closure, plus gives us the benefit of cancellation propagation.

In order the logic does the following:
- line 18 creates the target and holds a strong reference to it
- line 19 passes the target into the Picasso `into` method, which holds a soft reference
- line 20 adds a closure with a strong reference to target onto the `cont` (`Continuation`) object.
- once line 20 executes, the strong reference at line 18 is released.

- Sometime later the background logic in Picasso finishes
- This calls the target (which still has a strong reference) which will resumes the `Continuation` object
- The continuation object would then discard the unnecessary cancellation closure
- This would release the last strong reference to the target, which then allows it to be garbage collected